Implementing FileIntegrity structs

Simon9991 opened the pull request:
## Description
Basic boilerplating to create the `FileIntegrity` and `GameFile` structs.

Simon9991 committed:
wip(file_integrity): started to implement some structs for the files integrity

Simon9991 committed:
wip-fix(file_integrity): fixing the compile error

onkoe commented on src/file_integrity.rs:
```diff
@@ -0,0 +1,46 @@
+
+/// Fictional struct for the moment / prob use an external library
+#[derive(PartialEq)]
+pub struct Checksum {
```
*: Most types will want `Clone`, `Debug`, `PartialEq`, and `PartialOrd`

`Hash` is kinda optional, but I tend to add it in case I use `HashSet` over `Vec` somewhere


```suggestion
#[derive(Clone, Debug, PartialEq, PartialOrd, Hash)]
```

onkoe commented on src/file_integrity.rs:
```diff
@@ -0,0 +1,46 @@
+
+/// A MK8 game file and its expected `Checksum` (ref. value).
+#[derive(PartialEq)]
+struct GameFile {
```
see *

```suggestion
#[derive(Clone, Debug, PartialEq, PartialOrd, Hash)]
```

onkoe commented on src/file_integrity.rs:
```diff
@@ -0,0 +1,46 @@
+    }
+}
+
+/// Are all those derive useful?
```
note: This is a doc comment, so it renders into the documentation. You probably don't want that, so use two slashes: https://imgur.com/a/fiMEjpp

And, no, they're not currently useful. Like I said before, though, you'll want most of your types to implement traits "all the way down." A type can't be `Clone` unless its children are, nor can it be `PartialEq` or `Deserialize`. Which looks like a scary error until you realize... 😄️

Simon9991 committed:
fix(file_integrity): removing errors and warnings

Simon9991 committed:
feat(FileIntegrity): retrieving the game files

Simon9991 committed:
feat(sha2): added `sha2` crate

Simon9991 committed:
feat(FileIntegrity): placeholder to compute one checksum (test)

Simon9991 committed:
fix(FileIntegrity): return false if a file is not matching expected checksum

onkoe committed:
feat: Enable logging

onkoe committed:
feat: Add a configuration parsing error

onkoe committed:
refactor(config): Use `Result` over `Option`

onkoe commented:
uhh accidentally overwrote the actual parsing part with my lil refactor

will fix tmrw 🥹

onkoe committed:
fix: Scan files

whoops sorry simon

onkoe committed:
chore: Remove unused `print` methods

These are accessible using the `dbg!` macro for any type that
implements `Debug`, so creating them manually is unnecessary.

onkoe committed:
fix(file_integrity): Lots of string reallocations

If you use `format!`, a new String is allocated. Doing so in a loop
causes constant reallocations until the action is complete. Instead,
adding onto a single string will generally work out better in
terms of performance.

Clippy has a lint about exactly this, by the way!

https://rust-lang.github.io/rust-clippy/master/index.html#/format_collect

Make sure your IDE has `rust-analyzer` enabled to see these useful
lints.

onkoe committed:
feat(file_integrity): Use custom error type

this should have other useful types later on. no need to just
wrap std::io::Error :p

Simon9991 committed:
chore(FileIntegrity): removed unused import
